pkg/handlers/generic/lifecycle/csi/handler.go

```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		log.Error(
+			err,
+			"failed to read CSI provider from cluster definition",
+		)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(
+			fmt.Sprintf("failed to read CSI provider from cluster definition: %v",
+				err,
+			),
+		)
+		return
+	}
+	if !found || csiProviders.Providers == nil || len(csiProviders.Providers) == 0 {
+		log.V(4).Info(
+			fmt.Sprintf(
+				"Skipping CSI handler, no providers given in %v",
+				csiProviders,
+			),
+		)
+		return
+	}
+	for _, provider := range csiProviders.Providers {
+		handler, ok := c.ProviderHandler[provider.Name]
+		if !ok {
+			log.V(4).Info(
+				fmt.Sprintf(
+					"Skipping CSI handler, for provider given in %q. Provider handler not given ",
+					provider,
+				),
+			)
+			continue
+		}
+		cm, err := handler.EnsureCSIConfigMapForCluster(ctx, &req.Cluster)
+		if err != nil {
+			log.Error(
+				err,
+				fmt.Sprintf(
+					"failed to ensure %s csi driver installation manifests ConfigMap",
+					provider.Name,
+				),
+			)
+			resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		}
+		if cm != nil {
+			if provider.Name == csiProviders.DefaultClassName {
+				err = setDefaultStorageClass(ctx, cm)
+				if err != nil {
+					log.Error(
+						err,
+						fmt.Sprintf(
+							"failed to set default storage class",
+						),
+					)
+					resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+				}
+			}
+			err = c.EnsureCSICRSForCluster(ctx, &req.Cluster, cm)
+			if err != nil {
+				log.Error(
+					err,
+					fmt.Sprintf(
+						"failed to ensure %s csi driver installation manifests ConfigMap",
+						provider.Name,
+					),
+				)
+				resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+			}
+		}
+	}
+}
+
+func (c *CSIHandler) EnsureCSICRSForCluster(
+	ctx context.Context,
+	cluster *clusterv1.Cluster,
+	cm *corev1.ConfigMap,
+) error {
+	crs := &crsv1.ClusterResourceSet{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: crsv1.GroupVersion.String(),
+			Kind:       "ClusterResourceSet",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      cm.Name + "-" + cluster.Name,
+		},
+		Spec: crsv1.ClusterResourceSetSpec{
+			Resources: []crsv1.ResourceRef{{
+				Kind: string(crsv1.ConfigMapClusterResourceSetResourceKind),
+				Name: cm.Name,
+			}},
+			Strategy: string(crsv1.ClusterResourceSetStrategyReconcile),
+			ClusterSelector: metav1.LabelSelector{
+				MatchLabels: map[string]string{clusterv1.ClusterNameLabel: cluster.Name},
+			},
+		},
+	}
+
+	if err := controllerutil.SetOwnerReference(cluster, crs, c.client.Scheme()); err != nil {
+		return fmt.Errorf("failed to set owner reference: %w", err)
+	}
+	err := client.ServerSideApply(ctx, c.client, crs)
+	if err != nil {
+		return fmt.Errorf("failed to server side apply %w", err)
+	}
+	return nil
+}
+
+func setDefaultStorageClass(
+	ctx context.Context,
+	cm *corev1.ConfigMap,
+) error {
+	for k, contents := range cm.Data {
+		if strings.Contains(k, ".yaml") {
```

Comment:
I wonder if we should iterate through all the keys until we find a `kindStorageClass`, relying on just the `.yaml` may not be enpugh.
And then return an error if we don't find one to change.

Reply:
thats not a bad idea. it would be a lot safer to do it that way and allow us some flexibility as well.

Reply:
implemented